runtime-extension-model/src/main/java/org/mule/runtime/core/internal/extension/ForEachCollectionTypeResolver.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2023 Salesforce, Inc. All rights reserved.
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.internal.extension;
+
+import static org.mule.runtime.api.metadata.resolving.FailureCode.UNKNOWN;
+import static org.mule.runtime.core.api.extension.provider.MuleExtensionModelProvider.ANY_TYPE;
+
+import static java.lang.String.format;
+
+import org.mule.metadata.api.model.ArrayType;
+import org.mule.metadata.api.model.MetadataType;
+import org.mule.metadata.message.api.el.TypeBindings;
+import org.mule.runtime.api.connection.ConnectionException;
+import org.mule.runtime.api.metadata.ExpressionLanguageMetadataService;
+import org.mule.runtime.api.metadata.MetadataContext;
+import org.mule.runtime.api.metadata.MetadataResolvingException;
+import org.mule.runtime.api.metadata.resolving.InputTypeResolver;
+import org.mule.runtime.privileged.metadata.InternalMetadataContext;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * {@link InputTypeResolver} implementation used for the "collection" parameter of the ForEach and ParallelForEach scopes.
+ *
+ * @since 4.8.0
+ */
+public class ForEachCollectionTypeResolver implements InputTypeResolver<String> {
+
+  @Override
+  public String getCategoryName() {
+    return "FOREACH";
+  }
+
+  @Override
+  public String getResolverName() {
+    return "FOREACH_INPUT_COLLECTION";
+  }
+
+  @Override
+  public MetadataType getInputMetadata(MetadataContext context, String key)
+      throws MetadataResolvingException, ConnectionException {
+    if (context instanceof InternalMetadataContext) {
+      InternalMetadataContext internalContext = (InternalMetadataContext) context;
+      if (internalContext.getExpressionLanguageMetadataService().isPresent() && internalContext.getTypeBindings().isPresent()) {
```

Comment:
same.. fail if not

Reply:
The `TypeBindings` will only be available when doing propagation. If this is executed as part of plain type resolution, it is very possible that there are no `TypeBindings` provided (they can be provided, but they probably won't). I think falling back to ANY makes sense there.